controllers/drain_controller.go

```diff
@@ -0,0 +1,492 @@
+				"DrainController",
+				"node drain operation was not completed")
+			return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+
+		// if we manage to drain we label the node state with drain completed and finish
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainComplete, dr.Client)
+		if err != nil {
+			reqLogger.Error(err, "failed to annotate node with annotation", "annotation", constants.DrainComplete)
+			return ctrl.Result{}, err
+		}
+
+		reqLogger.Info("node drained successfully")
+		dr.recorder.Event(nodeNetworkState,
+			corev1.EventTypeWarning,
+			"DrainController",
+			"node drain completed")
+		return ctrl.Result{}, nil
+	}
+
+	reqLogger.Error(nil, "unexpected node drain annotation")
+	return reconcile.Result{}, fmt.Errorf("unexpected node drain annotation")
+}
+
+func (dr *DrainReconcile) checkForNodeDrain(ctx context.Context, node *corev1.Node) (*reconcile.Result, error) {
+	// configure logs
+	reqLogger := log.FromContext(ctx)
+	reqLogger.Info("checkForNodeDrain():")
+
+	//critical section we need to check if we can start the draining
+	dr.drainCheckMutex.Lock()
+	defer dr.drainCheckMutex.Unlock()
+
+	// find the relevant node pool
+	nodePool, nodeList, err := dr.findNodePoolConfig(ctx, node)
+	if err != nil {
+		reqLogger.Error(err, "failed to find the pool for the requested node")
+		return nil, err
+	}
+
+	// check how many nodes we can drain in parallel for the specific pool
+	maxUnv, err := nodePool.MaxUnavailable(len(nodeList))
+	if err != nil {
+		reqLogger.Error(err, "failed to calculate max unavailable")
+		return nil, err
+	}
+
+	current := 0
+	snns := &sriovnetworkv1.SriovNetworkNodeState{}
+
+	var currentSnns *sriovnetworkv1.SriovNetworkNodeState
+	for _, nodeObj := range nodeList {
+		err = dr.Get(ctx, client.ObjectKey{Name: nodeObj.GetName(), Namespace: vars.Namespace}, snns)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				reqLogger.V(2).Info("node doesn't have a sriovNetworkNodePolicy")
+				continue
+			}
+			return nil, err
+		}
+
+		if snns.GetName() == node.GetName() {
+			currentSnns = snns.DeepCopy()
+		}
+
+		if utils.ObjectHasAnnotation(snns, constants.NodeStateDrainAnnotationCurrent, constants.Draining) ||
+			utils.ObjectHasAnnotation(snns, constants.NodeStateDrainAnnotationCurrent, constants.DrainComplete) {
+			current++
+		}
+	}
+	reqLogger.Info("Max node allowed to be draining at the same time", "MaxParallelNodeConfiguration", maxUnv)
+	reqLogger.Info("Count of draining", "drainingNodes", current)
+
+	// if maxUnv is zero this means we drain all the nodes in parallel without a limit
+	if maxUnv == -1 {
+		reqLogger.Info("draining all the nodes in parallel")
+	} else if current >= maxUnv {
+		// the node requested to be drained, but we are at the limit so we re-enqueue the request
+		reqLogger.Info("MaxParallelNodeConfiguration limit reached for draining nodes re-enqueue the request")
+		// TODO: make this time configurable
```

Comment:
this ansers the question above. Please, add TODO note on line 221 too

Reply:
sure